src/Npgsql/ConnectorPool.cs

```diff
@@ -197,200 +190,172 @@ internal async ValueTask<NpgsqlConnector> AllocateLong(NpgsqlConnection conn, Np
         {
             // No idle connector was found in the pool.
             // We now loop until one of three things happen:
-            // 1. The pool isn't at max capacity (Total < Max), so we can create a new physical connection.
-            // 2. The pool is at maximum capacity and there are no idle connectors (Busy == Max),
+            // 1. The pool isn't at max capacity (Open < Max), so we can create a new physical connection.
+            // 2. The pool is at maximum capacity and there are no idle connectors (Open - Idle == Max),
             // so we enqueue an open attempt into the waiting queue, so that the next release will unblock it.
-            // 3. An connector makes it into the idle list (race condition with another Release().
+            // 3. An connector makes it into the idle list (race condition with another Release()).
             while (true)
             {
                 NpgsqlConnector? connector;
-                var state = State.Copy();
-                var newState = state;
 
-                if (state.Total < _max)
+                var (openCount, idleCount, busyCount) = State;
+                if (openCount < _max)
                 {
-                    // We're under the pool's max capacity, try to "allocate" a slot for a new physical connection.
-                    newState.Busy++;
-                    CheckInvariants(newState);
-                    if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) != state.All)
+                    // We're under the pool's max capacity, "allocate" a slot for a new physical connection.
+                    // Don't spin for this https://github.com/dotnet/coreclr/pull/21437
+                    var prevOpenCount = openCount;
+                    while (true)
                     {
-                        // Our attempt to increment the busy count failed, Loop again and retry.
-                        continue;
+                        var currentOpenCount = prevOpenCount;
+                        prevOpenCount = Interlocked.CompareExchange(ref State.Open, currentOpenCount + 1, currentOpenCount);
+                        // Either we succeeded or someone else did and we're at max opens, break.
+                        if (prevOpenCount == currentOpenCount || prevOpenCount == _max) break;
```

Comment:
Thanks for fixing this. It may be a bit more readable to use goto to jump out rather than checking _max twice - up to you.

Otherwise, as a nit, in general we put `break` on a separate line (but no curlies required). Same for `continue` below.

Reply:
Wasn't sure if you were ok with GOTO, I actually stopped myself a few times already, hah.

Reply:
Also the initial implementation (with break on the same line) comes directly from @YohDeadfall ^_^ 

https://github.com/npgsql/npgsql/pull/2519#discussion_r304079135